Nested (left) join regression

**Version**: `>=1.0.0`
**Module**: `quill-sql`
**Database**: `-`

### Expected behavior

```sql
SELECT x02.link, x01.id, x11.link FROM a x02 LEFT JOIN b x01 INNER JOIN c x11 ON x11.link = x01.id ON x02.link = x01.id
```

### Actual behavior

I believe #600 might be related.

```shell
ScalaFiddle.scala:11: error: exception during macro expansion: 
scala.reflect.macros.TypecheckException: value flatMap is not a member of ScalaFiddle.C
	at scala.reflect.macros.contexts.Typers$$anonfun$typecheck$2$$anonfun$apply$1.apply(Typers.scala:34)
	at scala.reflect.macros.contexts.Typers$$anonfun$typecheck$2$$anonfun$apply$1.apply(Typers.scala:28)
	at scala.reflect.macros.contexts.Typers$$anonfun$3.apply(Typers.scala:24)
```
### Steps to reproduce the behavior

```scala
val ctx = new SqlMirrorContext[PostgresDialect, SnakeCase]
import ctx._

case class A(link: Int)
case class B(id: Int)
case class C(link: Int)

val m = ctx.run {
  query[A]
    .leftJoin {
      query[B]
        .join(query[C])
        .on { case (b, c) => c.link == b.id }
    }
    .on { case (a, (b, c)) => a.link == b.id }
}
```

ScalaFiddle available [here](https://scalafiddle.io/sf/6Iy2Yv0/3). It seems to work with version below `1.0.0` [here](https://scalafiddle.io/sf/6Iy2Yv0/1). ~~Right join does not seem to have the same issue.~~ Also it is not possible to use the latest quill version on the fiddle [here](https://scalafiddle.io/sf/6Iy2Yv0/2).

### Workaround

Might be rewritten another way around however less easy to use.

@getquill/maintainers

--- a/quill-core/src/main/scala/io/getquill/dsl/MetaDslMacro.scala
+++ b/quill-core/src/main/scala/io/getquill/dsl/MetaDslMacro.scala
@@ -54,16 +54,15 @@ class MetaDslMacro(val c: MacroContext) {
     actionMeta[T](value("Encoder", t.tpe), "insert")
 
   def materializeSchemaMeta[T](implicit t: WeakTypeTag[T]): Tree =
-    (t.tpe.typeSymbol.isClass && t.tpe.typeSymbol.asClass.isCaseClass) match {
-      case true =>
-        q"""
+    if (t.tpe.typeSymbol.isClass && t.tpe.typeSymbol.asClass.isCaseClass) {
+      q"""
           new ${c.prefix}.SchemaMeta[$t] {
             val entity =
               ${c.prefix}.quote(${c.prefix}.querySchema[$t](${t.tpe.typeSymbol.name.decodedName.toString}))
           }
         """
-      case false =>
-        c.fail(s"Can't materialize a `SchemaMeta` for non-case-class type '${t.tpe}', please provide an implicit `SchemaMeta`.")
+    } else {
+      c.fail(s"Can't materialize a `SchemaMeta` for non-case-class type '${t.tpe}', please provide an implicit `SchemaMeta`.")
     }
 
   private def expandQuery[T](value: Value)(implicit t: WeakTypeTag[T]) = {
@@ -73,38 +72,41 @@ class MetaDslMacro(val c: MacroContext) {
 
   private def extract[T](value: Value)(implicit t: WeakTypeTag[T]): Tree = {
     var index = -1
-    def expand(value: Value, optional: Boolean = false): Tree =
+
+    def expand(value: Value, parentOptional: Boolean = false): Tree =
       value match {
 
-        case Scalar(path, tpe, decoder) =>
+        case Scalar(_, tpe, decoder, optional) =>
           index += 1
-          optional match {
-            case true =>
+          if (parentOptional) {
+            if (optional)
+              q"Some($decoder($index, row))"
+            else
               q"implicitly[${c.prefix}.Decoder[Option[$tpe]]].apply($index, row)"
-            case false =>
-              q"$decoder($index, row)"
+          } else {
+            q"$decoder($index, row)"
           }
 
-        case Nested(term, tpe, params) =>
-          q"new $tpe(...${params.map(_.map(expand(_)))})"
-
-        case OptionalNested(term, tpe, params) =>
-          val groups = params.map(_.map(expand(_, optional = true)))
-          val terms =
-            groups.zipWithIndex.map {
-              case (options, idx1) =>
-                (0 until options.size).map { idx2 =>
-                  TermName(s"o_${idx1}_${idx2}")
+        case Nested(_, tpe, params, optional) =>
+          if (parentOptional || optional) {
+            val groups = params.map(_.map(expand(_, parentOptional = true)))
+            val terms =
+              groups.zipWithIndex.map {
+                case (options, idx1) =>
+                  options.indices.map { idx2 =>
+                    TermName(s"o_${idx1}_$idx2")
+                  }
+              }
+            groups.zipWithIndex.foldLeft(q"Some(new $tpe(...$terms))") {
+              case (body, (options, idx1)) =>
+                options.zipWithIndex.foldLeft(body) {
+                  case (body, (option, idx2)) =>
+                    val o = q"val ${TermName(s"o_${idx1}_$idx2")} = $EmptyTree"
+                    q"$option.flatMap($o => $body)"
                 }
             }
-          groups.zipWithIndex.foldLeft(q"Some(new $tpe(...$terms))") {
-            case (body, (options, idx1)) =>
-              options.zipWithIndex.foldLeft(body) {
-                case (body, (option, idx2)) =>
-                  val o = q"val ${TermName(s"o_${idx1}_${idx2}")} = $EmptyTree"
-                  q"$option.flatMap($o => $body)"
-              }
-          }
+          } else
+            q"new $tpe(...${params.map(_.map(expand(_)))})"
       }
     q"(row: ${c.prefix}.ResultRow) => ${expand(value)}"
   }
@@ -133,25 +135,27 @@ class MetaDslMacro(val c: MacroContext) {
         case None       => tree
         case Some(term) => q"$tree.$term"
       }
+
     def apply(base: Tree, params: List[List[Value]]): List[Tree] =
-      params.flatten.map(flatten(base, _)).flatten
+      params.flatten.flatMap(flatten(base, _))
+
     value match {
-      case Scalar(term, tpe, decoder) =>
+      case Scalar(term, _, _, _) =>
         List(nest(base, term))
-      case Nested(term, tpe, params) =>
-        apply(nest(base, term), params)
-      case OptionalNested(term, tpe, params) =>
-        apply(q"v", params)
-          .map(body => q"${nest(base, term)}.map(v => $body)")
+      case Nested(term, _, params, optional) =>
+        if (optional)
+          apply(q"v", params)
+            .map(body => q"${nest(base, term)}.map(v => $body)")
+        else
+          apply(nest(base, term), params)
     }
   }
 
   sealed trait Value {
     val term: Option[TermName]
   }
-  case class Nested(term: Option[TermName], tpe: Type, params: List[List[Value]]) extends Value
-  case class OptionalNested(term: Option[TermName], tpe: Type, params: List[List[Value]]) extends Value
-  case class Scalar(term: Option[TermName], tpe: Type, decoder: Tree) extends Value
+  case class Nested(term: Option[TermName], tpe: Type, params: List[List[Value]], optional: Boolean) extends Value
+  case class Scalar(term: Option[TermName], tpe: Type, decoder: Tree, optional: Boolean) extends Value
 
   private def is[T](tpe: Type)(implicit t: TypeTag[T]) =
     tpe <:< t.tpe
@@ -173,12 +177,13 @@ class MetaDslMacro(val c: MacroContext) {
                 )
               }
             }
-          Nested(term, tpe, params)
+          Nested(term, tpe, params, optional = false)
       }
 
     def apply(tpe: Type, term: Option[TermName], nested: Boolean): Value = {
       OptionalTypecheck(c)(q"implicitly[${c.prefix}.${TypeName(encoding)}[$tpe]]") match {
-        case Some(encoding) => Scalar(term, tpe, encoding)
+        case Some(encoding) =>
+          Scalar(term, tpe, encoding, optional = is[Option[Any]](tpe))
         case None =>
 
           def value(tpe: Type) =
@@ -193,12 +198,10 @@ class MetaDslMacro(val c: MacroContext) {
                 nest(tpe, term)
             }
 
-          is[Option[Any]](tpe) match {
-            case false =>
-              value(tpe)
-            case true =>
-              val nested = value(tpe.typeArgs.head)
-              OptionalNested(nested.term, nested.tpe, nested.params)
+          if (is[Option[Any]](tpe)) {
+            value(tpe.typeArgs.head).copy(optional = true)
+          } else {
+            value(tpe)
           }
       }
     }
@@ -211,19 +214,17 @@ class MetaDslMacro(val c: MacroContext) {
               tree match {
                 case q"$a.$b"                => path(a) :+ b
                 case q"$a.map[$t]($b => $c)" => path(a) ++ path(c)
-                case other                   => Nil
+                case _                       => Nil
               }
             path(f.body)
         }
 
       def filter(value: Value, path: List[TermName] = Nil): Option[Value] =
         value match {
-          case value if (paths.contains(path ++ value.term)) =>
+          case value if paths.contains(path ++ value.term) =>
             None
-          case Nested(term, tpe, params) =>
-            Some(Nested(term, tpe, params.map(_.map(filter(_, path ++ term)).flatten)))
-          case OptionalNested(term, tpe, params) =>
-            Some(OptionalNested(term, tpe, params.map(_.map(filter(_, path ++ term)).flatten)))
+          case Nested(term, tpe, params, optional) =>
+            Some(Nested(term, tpe, params.map(_.flatMap(filter(_, path ++ term))), optional))
           case value =>
             Some(value)
         }
@@ -241,6 +242,6 @@ class MetaDslMacro(val c: MacroContext) {
 
   private def caseClassConstructor(t: Type) =
     t.members.collect {
-      case m: MethodSymbol if (m.isPrimaryConstructor) => m
+      case m: MethodSymbol if m.isPrimaryConstructor => m
     }.headOption
 }

--- a/quill-core/src/test/scala/io/getquill/dsl/MetaDslSpec.scala
+++ b/quill-core/src/test/scala/io/getquill/dsl/MetaDslSpec.scala
@@ -70,10 +70,30 @@ class MetaDslSpec extends Spec {
           meta.extract(Row("a", Some("1"), None)) mustEqual (("a", None))
         }
       }
+      "optional deep nested" - {
+        case class Entity1(a: String, b: Int)
+        case class Entity2(a: Option[Int])
+
+        val meta = materializeQueryMeta[(String, Option[(Entity1, Entity2)])]
+        meta.expand.toString mustEqual "(q) => q.map(x => (x._1, x._2.map((v) => v._1.a), x._2.map((v) => v._1.b), x._2.map((v) => v._2.a)))"
+
+        "extracts Some if all columns are defined" in {
+          meta.extract(Row("a", Some("1"), Some(2), Some(3))) mustEqual
+            (("a", Some((Entity1("1", 2), Entity2(Some(3))))))
+        }
+        "extracts Some if optional column is undefined" in {
+          meta.extract(Row("a", Some("1"), Some(2), None)) mustEqual
+            (("a", Some((Entity1("1", 2), Entity2(None)))))
+        }
+        "extracts None if one column is undefined" in {
+          meta.extract(Row("a", Some("1"), None, Some(3))) mustEqual
+            (("a", None))
+        }
+      }
       "> 22 fields" in {
         val meta = materializeQueryMeta[MoreThan22]
         meta.expand.toString mustEqual "(q) => q.map(x => (x.v0, x.v1, x.v2, x.v3, x.v4, x.v5, x.v6, x.v7, x.v8, x.v9, x.x0, x.x1, x.x2, x.x3, x.x4, x.x5, x.x6, x.x7, x.x8, x.x9, x.y0, x.y1, x.y2, x.y3, x.y4, x.y5, x.y6, x.y7, x.y8, x.y9))"
-        meta.extract(Row((0 until 30): _*)) mustEqual
+        meta.extract(Row(0 until 30: _*)) mustEqual
           MoreThan22(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29)
       }
     }

